### ISSUE TITLE
Possible bugs in YOLO Layer's forward

### ISSUE THREAD

**Issue by alexandru-dinu** [eriklindernoren/PyTorch-YOLOv3#33]
1. The _no object loss_ component of the confidence loss:
```
self.lambda_noobj * self.bce_loss(conf * (1 - mask), mask * (1 - mask))
```
basically compares `conf * (1 - mask)` with `mask * (1 - mask)`, but `mask * (1 - mask)` will be a tensor full of zeros. I think it should be:
```
self.lambda_noobj * self.bce_loss(conf * (1 - mask), tconf * (1 - mask))
```

2. The AP calculation has 2 problems: 

- isn't `nCorrect / nGT` actually the recall? precision is the number of correct predictions divided by the number of all predictions.
- if there are no ground truths, `nCorrect / nGT` will be `nan`, so the correct expression must be:
`1 if (nCorrect == nGT == 0) else (nCorrect / nGT)`

**Comment by PiggyGenius** [comment]
I think you are right on both points. I modified my code as you suggested but haven't tried training yet.

**Comment by wei2ran** [comment]
tconf and mask are defined same in utils

**Comment by PiggyGenius** [comment]
Yes we can actually remove tconf and just use ```(1 - mask)```

**Comment by PiggyGenius** [comment]
Actually removing ```mask * (1 - mask)``` is a problem. You want the noobject conf target to be made of zeros since you want them to have a low confidence. The way it was is correct I think.

**Comment by eriklindernoren** [comment]
This should be fixed in 1dcdf11
